programs/loader-v3/src/lib.rs

```diff
@@ -0,0 +1,1369 @@
+    let instruction_data = instruction_context.get_instruction_data();
+    let program_id = instruction_context.get_last_program_key(transaction_context)?;
+    if loader_v3::check_id(program_id) {
+        match limited_deserialize(instruction_data)? {
+            LoaderV3Instruction::Write { offset, bytes } => {
+                let mut program =
+                    instruction_context.try_borrow_instruction_account(transaction_context, 0)?;
+                let authority_address = transaction_context.get_key_of_account_at_index(
+                    instruction_context.get_index_of_instruction_account_in_transaction(1)?,
+                )?;
+                let mut payer = instruction_context
+                    .try_borrow_instruction_account(transaction_context, 2)
+                    .ok();
+                let is_initialization = offset == 0 && program.get_data().is_empty();
+                if is_initialization {
+                    if !loader_v3::check_id(program.get_owner()) {
+                        ic_logger_msg!(log_collector, "Program not owned by loader");
+                        return Err(InstructionError::InvalidAccountOwner);
+                    }
+                    if !program.is_writable() {
+                        ic_logger_msg!(log_collector, "Program is not writeable");
+                        return Err(InstructionError::InvalidArgument);
+                    }
+                    if !instruction_context.is_instruction_account_signer(1)? {
+                        ic_logger_msg!(log_collector, "Authority did not sign");
+                        return Err(InstructionError::MissingRequiredSignature);
+                    }
+                } else {
+                    let state = check_program_account(
+                        &log_collector,
+                        instruction_context,
+                        &program,
+                        authority_address,
+                    )?;
+                    if state.is_deployed {
+                        ic_logger_msg!(log_collector, "Program is not retracted");
+                        return Err(InstructionError::InvalidArgument);
+                    }
+                }
+                if payer.is_some() && !instruction_context.is_instruction_account_signer(2)? {
+                    ic_logger_msg!(log_collector, "Payer did not sign");
+                    return Err(InstructionError::MissingRequiredSignature);
+                }
+                if payer.is_some() && !instruction_context.is_instruction_account_writable(2)? {
+                    ic_logger_msg!(log_collector, "Payer is not writeable");
+                    return Err(InstructionError::InvalidArgument);
+                }
+                let program_size = program
+                    .get_data()
+                    .len()
+                    .saturating_sub(LoaderV3State::program_data_offset());
+                if offset as usize > program_size {
```

Comment:
Should offset be = program_size? Or do we allow sparse file to store the program? 
Or maybe it is to support out of order writes. But in that case the write to fill in the gap will fail this check (as the program_size will be > offset).

Reply:
Oh, I think I misread the code. Ignore this comment.